### PR TITLE
automatically rotate bottom camera for switchable cameras

### PR DESCRIPTION
**Pull request by e3l**

### CODE REVIEW

File: src/main/java/frc/robot/vision/SwitchableCamera.java [hunk 1]
```diff
@@ -1,40 +1,90 @@
...
+
+    private final int WIDTH = 140;
+    private final int HEIGHT = 120;
+
+    private final CvSink bottomSink;
+    private final CvSink topSink;
+
+    private final CvSource server;
+
+    private volatile boolean frontActive = true;
```
Comment (ky28059) [inline src/main/java/frc/robot/vision/SwitchableCamera.java:28]: `volatile` 😳

File: src/main/java/frc/robot/vision/SwitchableCamera.java [hunk 2]
```diff
@@ -1,40 +1,90 @@
...
+        topCam.setResolution(WIDTH, HEIGHT);
+        bottomCam.setResolution(WIDTH, HEIGHT);
+
+        topCam.setFPS(30);
+        bottomCam.setFPS(30);
+
+        bottomSink = new CvSink("flip bottom cam");
+        bottomSink.setSource(bottomCam);
+
+        topSink = new CvSink("flip bottom cam");
```
Comment (ky28059) [inline src/main/java/frc/robot/vision/SwitchableCamera.java:43]: Are both of these supposed to have the same name?

File: src/main/java/frc/robot/vision/SwitchableCamera.java [hunk 3]
```diff
@@ -1,40 +1,90 @@
...
+        topSink.setSource(topCam);
+
+        this.server = CameraServer.putVideo("test", HEIGHT, WIDTH);
```
Comment (ky28059) [inline src/main/java/frc/robot/vision/SwitchableCamera.java:46]: `"test"`

File: src/main/java/frc/robot/vision/SwitchableCamera.java [hunk 4]
```diff
@@ -1,40 +1,90 @@
...
+            while (!Thread.interrupted()) {
+                if (frontActive) {
+                    if (bottomSink.grabFrame(img) != 0) {
+                        Core.rotate(img, img, Core.ROTATE_90_CLOCKWISE);
+                        server.putFrame(img);
+                    } else {
+                        DriverStation.reportError(bottomSink.getError(), false);
+                    }
+                } else {
+                    if (topSink.grabFrame(img) != 0) {
```
Comment (ky28059) [inline src/main/java/frc/robot/vision/SwitchableCamera.java:59]: Maybe the `grabFrame` logic stuff can be abstracted into a method with a boolean for whether to rotate it?

File: src/main/java/frc/robot/vision/SwitchableCamera.java [hunk 5]
```diff
@@ -1,40 +1,90 @@
...
  * A `VideoSource` that toggles between a front and back camera.
  */
 public class SwitchableCamera {
-    private final UsbCamera front;
-    private final UsbCamera back;
-    private final VideoSink server;
+    private final UsbCamera topCam;
+    private final UsbCamera bottomCam;
+
+    private final int WIDTH = 140;
```
Comment (ky28059) [inline src/main/java/frc/robot/vision/SwitchableCamera.java:20]: Doesn't really matter but could be static.

File: src/main/java/frc/robot/vision/SwitchableCamera.java (hunk 2 above)
Comment (e3l) [reply src/main/java/frc/robot/vision/SwitchableCamera.java:43]: no

**Comment by e3l** [comment]
showed significantly higher latency

**Pull request closed without merging**